File: src/test/java/fi/aalto/cs/apluscourses/utils/VersionTest.java
package fi.aalto.cs.apluscourses.utils;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotEquals;
import static org.junit.Assert.fail;
import static org.junit.jupiter.api.Assertions.assertThrows;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

class VersionTest {

  @Test
  void testCreateVersion() {
    Version version = new Version(3, 5);

    Assertions.assertEquals(3, version.major, "The major version should be the same as that given to the constructor");
    Assertions.assertEquals(5, version.minor, "The minor version should be the same as that given to the constructor");

    Assertions.assertEquals("3.5", version.toString(),
        "toString() should return the version in format '{major}.{minor}'.");

    Version sameVersion = new Version(3, 5);
    Assertions.assertEquals(version, sameVersion,
        "Version should equal to another version created with the same arguments.");
    Assertions.assertEquals(version.hashCode(), sameVersion.hashCode(),
        "Two equal versions should give the same hash code.");

    Version differentMajor = new Version(3, 6);
    Assertions.assertNotEquals(version, differentMajor, "Versions with different majors versions should not be equal");

    Version differentMinor = new Version(4, 5);
    Assertions.assertNotEquals(version, differentMinor, "Versions with different minor versions should not be equal");
  }

  @Test
  void testCreateVersionWithNegativeMajorVersion() {
    assertThrows(IllegalArgumentException.class, () ->
        new Version(-1, 4));
  }

  @Test
  void testCreateVersionWithNegativeMinorVersion() {
    assertThrows(IllegalArgumentException.class, () ->
        new Version(3, -13));
  }

  @Test
  void testEmptyVersion() {
    Version version = Version.EMPTY;

    Assertions.assertEquals(0, version.major);
    Assertions.assertEquals(0, version.minor);
  }

  @Test
  void testCreateVersionFromValidString() {
    Version version = Version.fromString("14.2");

    Assertions.assertEquals(14, version.major, "Version created by fromString(\"14.2\") should have major version 4.");
    Assertions.assertEquals(2, version.minor, "Version created by fromString(\"14.2\") should have minor version 2.");
  }

  @Test
  void testCreateVersionFromInvalidString() {
    String versionString = "invalid_version";

    try {
      Version.fromString(versionString);
    } catch (Version.InvalidVersionStringException ex) {
      Assertions.assertEquals(versionString, ex.getVersionString(),
          "Exception should have version string that was tried to parse.");
      return;
    }
    Assertions.fail("fromString() should throw an InvalidVersionStringException if the string is invalid.");
  }

  @Test
  void testCreateVersionFromStringMissingPart() {
    assertThrows(Version.InvalidVersionStringException.class, () ->
        Version.fromString("1"));
  }

  @Test
  void testCreateVersionFromStringMissingNumber() {
    assertThrows(Version.InvalidVersionStringException.class, () ->
        Version.fromString("9."));
  }

  @Test
  void testCreateVersionFromStringWithNonNumericPart() {
    assertThrows(Version.InvalidVersionStringException.class, () ->
        Version.fromString("6.XVII"));
  }

  @Test
  void testCreateVersionFromEmptyString() {
    assertThrows(Version.InvalidVersionStringException.class, () ->
        Version.fromString(""));
  }
}

File: src/test/kotlin/fi/aalto/cs/apluscourses/utils/VersionTest.kt
package fi.aalto.cs.apluscourses.utils

import fi.aalto.cs.apluscourses.utils.Version.ComparisonStatus
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

class VersionTest {
    @Test
    fun `parses major and minor from a version string`() {
        val version = Version("4.12")
        assertEquals(4, version.major)
        assertEquals(12, version.minor)
    }

    @Test
    fun `comparisonStatus reports how a version relates to the required one`() {
        assertEquals(ComparisonStatus.VALID, Version("4.2").comparisonStatus(Version("4.2")))
        assertEquals(ComparisonStatus.VALID, Version("4.3").comparisonStatus(Version("4.2")))
        assertEquals(ComparisonStatus.MINOR_TOO_OLD, Version("4.1").comparisonStatus(Version("4.2")))
        assertEquals(ComparisonStatus.MAJOR_TOO_OLD, Version("3.9").comparisonStatus(Version("4.0")))
        assertEquals(ComparisonStatus.MAJOR_TOO_NEW, Version("5.0").comparisonStatus(Version("4.9")))
    }

    @Test
    fun `equality and toString round-trip through the version string`() {
        assertEquals(Version("2.7"), Version(2, 7))
        assertEquals("2.7", Version(2, 7).toString())
    }

    @Test
    fun `the plugin version is packaged as a resource and parses`() {
        assertTrue(PluginVersion.current.isNotBlank(), "version resource missing from the plugin")
        assertTrue(
            PluginVersion.current.matches(Regex("""\d+\.\d+\.\d+.*""")),
            "unexpected version string: ${PluginVersion.current}"
        )
        assertEquals(PluginVersion.current, PluginVersion.currentVersion.toString())
    }
}
